Implement simple optimizations for predicates

tillarnold opened the pull request:
@vakaras this is the pull request for the optimizations we talked about in our meeting. I implemented the body removal idea we discussed as `remove_body_of_predicates_if_possible`. However it appears to remove predicate bodies that are actually needed so I commented out the lines where it is applied. I tried implementing a more naive approach in `remove_unnecessary_bodies` which does appear to work.

tillarnold committed:
Add initial version of delete unused predicates optimization

tillarnold committed:
Use ExprWalker and StmtWalker to implement unused predicate optimisation

tillarnold committed:
Clean up delete unused predicates optimisation

tillarnold committed:
Add initial version of delete not needed bodies optimisation

tillarnold committed:
Clean up predicate optimizations

tillarnold committed:
Take StructPredicates into account when computing used predicates

vakaras commented on prusti-common/src/vir/optimizations/mod.rs:
```diff
@@ -28,6 +29,17 @@ impl Program {
+        self.viper_predicates = predicates::delete_unused_predicates(
+            &self.methods,
+            &self.functions,
+            &self.viper_predicates,
```
I think it would be better to take the vector by value so that we can avoid a potentially expensive `to_vec()` inside the body of `delete_unused_predicates`.

vakaras commented on prusti-common/src/vir/optimizations/mod.rs:
```diff
@@ -28,6 +29,17 @@ impl Program {
+        self.viper_predicates = predicates::remove_unnecessary_bodies(
+            &self.methods,
+            &self.functions,
+            &self.viper_predicates,
```
Same here.

vakaras commented on prusti-common/src/vir/optimizations/predicates/mod.rs:
```diff
@@ -0,0 +1,49 @@
+};
+
+/// Walks all Statements and Expressions in the provided methods
+fn walk_methods(methods: &[CfgMethod], walker: &mut (impl StmtWalker + ExprWalker)) {
```
I think that it would be good to place these methods in VIR utils module or somewhere like that.

@fpoli @cmatheja @Aurel300  What do you think?

fpoli replied:
This looks generic enough to go in VIR :+1: (of course, trying to avoid `utils` as name of modules).

vakaras commented on prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs:
```diff
@@ -0,0 +1,191 @@
+    super::walk_methods(methods, &mut collector);
+    super::walk_functions(functions, &mut collector);
+
+    // DeadBorrowToken$ is a used predicate but it does not appear in VIR becaue it is only created when viper code is created from VIR
```
Would be good to keep the lines containing comments no more than 80 characters long.

vakaras commented on prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs:
```diff
@@ -0,0 +1,191 @@
+            Predicate::Struct(StructPredicate { body: Some(e), .. }) => {
+                ExprWalker::walk(&mut collector, e)
+            }
+            Predicate::Struct(_) => { /* ignore */ }
```
I think it would be better to make the code more explicit.
```suggestion
            Predicate::Struct(StructPredicate { body: None, .. }) => { /* ignore */ }
```

vakaras commented on prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs:
```diff
@@ -0,0 +1,191 @@
+        );
+        new_predicates = new_predicates
+            .into_iter()
+            .filter(|p| {
```
You probably could use [`Vec::retain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain) here.

vakaras commented on prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs:
```diff
@@ -0,0 +1,191 @@
+            .collect();
+    }
+
+    // FIXME: This acctually removes bodies that are needed
```
I would prefer to avoid the fixpoint if possible because it could void the gains of the optimization.

Have you checked where the predicates with removed bodies are used? Are they used only in `fold`, `unfold`, and `unfolding`? If yes, we could collect the predicates whose folding state is changed in a special list and remove only the bodies of predicates that are not mentioned in that list.

vakaras replied:
In other words: could the optimization implemented in `prusti-common/src/vir/optimizations/predicates/remove_unnecessary_bodies.rs` be merged into this file?

tillarnold replied:
I rearranged the optimisations such that remove_unnecessary_bodies is applied first. Even in that case it appears that around 5 of the tests in the test suite require more than 1 iteration to remove all unused predicates. Should I remove the fixpoint iteration since in most cases most unused predicates are removed that way or do we aim to really eliminate all unused predicates?

fpoli replied:
When you remove an unused predicates `P` you know precisely which other predicates might become unused: they are the predicates that were used by `P`. So, you can keep a much smaller set of predicates to be checked in the next iteration.

vakaras commented on prusti-common/src/vir/optimizations/predicates/delete_unused_predicates.rs:
```diff
@@ -0,0 +1,191 @@
+    while has_changed {
+        has_changed = false;
+
+        let predicates_used_in_predicates = get_used_predicates_in_predicates(&new_predicates);
```
I think it would be more performant to create a map `predicate_name → set of predicate names used in that predicate` that is computed once instead of recomputing this information over and over again in the loop.

vakaras reviewed (commented):
Thank you! The start looks very good. I have left some comments.

Aurel300 commented on prusti-common/src/vir/optimizations/predicates/mod.rs:
```diff
@@ -0,0 +1,49 @@
+    }
+}
+
+/// Walks all Expressions in the provided functions (including pre and post condidions)
```
```suggestion
/// Walks all Expressions in the provided functions (including pre and post conditions)
```

tillarnold committed:
Implement changes proposed in code review

tillarnold committed:
Implement delete_unused_predicates using a map

This way we only need to traverse the VIR only once and we compute
the reachable predicates by traversing the map

tillarnold committed:
Move walk_functions and walk_methods to utils module

tillarnold committed:
Remove dbg! statements

tillarnold committed:
Remove accidentally checked in file

tillarnold committed:
Remove old version of delete_unused_predicates

tillarnold committed:
Merge the delete_unused_predicates and remove_unnecessary_bodies optimisations

tillarnold committed:
Fix typo

vakaras reviewed (approved):
LGTM. However, we need to check why the tests are failing.
